### PR TITLE
Fixes backend unit tests and finishes code restructure

### PR DESCRIPTION
**Pull request by thinkharderdev**
## Highlights 
* Put DB interactions into service classes which are accessed through the ServiceContainer
* Mock authentication and DB interactions in unit tests
* Use Blueprints for modularizing handlers
* Add authenticate decorator to add auth to handler methods

### CODE REVIEW

File: backend/README.md [hunk 1]
```diff
@@ -7,9 +7,7 @@ RESTful API to handle storage, retrieval, and searching of landlords, navigators
 ## Setup
 
 - [install python 3.6.x](https://realpython.com/installing-python/)
-- [install npm](https://www.npmjs.com/get-npm)
-- [install the Serverless framework](https://serverless.com/framework/docs/providers/aws/guide/installation/)
-- [install the psql CLI](https://www.pgcli.com/install)
```
Comment (dgoldstein1) [inline backend/README.md:12]: why get rid of these links?

File: backend/README.md [hunk 2]
```diff
@@ -30,30 +28,22 @@ npm i
...
-export DB_HOST="localhost"
-export DB_PORT=5432
-export DB_USER="app"
-export DB_PASSWORD="apppassword"
-export APP_ENV="local"
+export APP_ENV=local
 $(cd alembic && alembic upgrade head)
 python3 api/swagger-yaml-to-html.py < api/swagger.yml > api/index.html
-flask run
+serverless wsgi serve
```
Comment (dgoldstein1) [inline backend/README.md:46]: this is failing for me

```bash
(venv) david@david-Oryx-Pro:~/dev/personal/housinghub/backend$ serverless wsgi serve
Serverless: Using Python specified in "runtime": python3.6
Unable to import werkzeug (run: pip install werkzeug)
(venv) david@david-Oryx-Pro:~/dev/personal/housinghub/backend$ npx serverless wsgi serve
Serverless: Using Python specified in "runtime": python3.6
Unable to import werkzeug (run: pip install werkzeug)
```

File: backend/api/swagger.yml [hunk 3]
```diff
@@ -454,35 +445,18 @@ components:
...
-          example: "david2134"
-        role:
-          type: string
-          description: what the user does in the system
-          example: "navigator"
-        is_admin:
-          type: boolean
-          description: is the user an admin in system?
-          example: true
+          description: UNDER CONSTRUCTION
```
Comment (dgoldstein1) [inline backend/api/swagger.yml:459]: why revert this change? It's nice to have the /auth endpoints fleshed out in the API now that we support them

File: backend/src/auth_handlers.py [hunk 4]
```diff
@@ -1,131 +1,63 @@
...
+        password = user_data.get('password')
+        role = user_data.get('role')
+        is_admin = bool(user_data.get('is_admin'))
+        if username and password and role and is_admin:
+            created_user = app.services.user_service().add_user(username,password,role,is_admin)
+            return jsonify({'jwt': app.services.auth_service().encode_jwt(created_user)})
+        else:
+            return jsonify(code=400, error='Request is invalid'), 400
+    except KeyError:
+        return jsonify(code=400, error='Request is invalid'), 400
```
Comment (dgoldstein1) [inline backend/src/auth_handlers.py:26]: errors should be logged here for debugging if a request returns a non-200 response

File: backend/src/auth_handlers.py [hunk 5]
```diff
@@ -1,131 +1,63 @@
...
-        return server.err_out(401, err)
-    # create a JWT token and return to front end
-    return jsonify({'jwt': encodeJWT(user)})
-
-
+    user = app.services.auth_service().validate_login(u, p)
+    if user:
+        # create a JWT token and return to front end
+        return jsonify({'jwt': app.services.auth_service().encode_jwt(user)})
+    else:
```
Comment (dgoldstein1) [inline backend/src/auth_handlers.py:48]:
```suggestion
```
don't need `else..return`, applies to other instances

File: backend/src/server.py [hunk 6]
```diff
@@ -1,174 +1,6 @@
...
-
-for endpt in supportedCrudEndpoints:
-    for m in endpt.get("methods"):
-        app.add_url_rule(endpt.get("path"),
-                         "{} a {}".format(m.get("method"), endpt.get("name")),
-                         _wrap_auth_handler(m.get("handler")),
-                         methods=[m.get("method")])
-
-# docs
-app.add_url_rule('/', "swagger docs", serve_docs)
```
Comment (dgoldstein1) [inline backend/src/server.py:171]: :confetti_ball: way cleaner!! Nice to delete this code

File: backend/watch_dev_changes.sh [hunk 7]
```diff
@@ -1,21 +1,14 @@
...
-
-python api/swagger-yaml-to-html.py < api/swagger.yml > api/index.html
 
 inotifywait -e modify,create,delete -r ./ && \
 	clear && \
-	yapf -ri ./src/*.py && \
-	autoflake --in-place --remove-unused-variables ./src/*.py && \
+	yapf -ri ./**/*.py && \
+	autoflake --in-place --remove-unused-variables ./**/*.py && \
 	coverage run -m pytest
```
Comment (dgoldstein1) [inline backend/watch_dev_changes.sh:13]: this needs to be ./src/* otherwise there is an error

```
    raise e
  File "./alembic/env.py", line 32
    return f'{app_env}.sqlalchemy.url'
```                                     ^

File: backend/src/app/auth.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+
+def authenticate(f):
+  @wraps(f)
+  def decorated_function(*args, **kwargs):
+    auth_service = app.services.auth_service()
+    (user, err) = auth_service.authenticate_request(request)
+    if user:
+      return f(*args,**kwargs)
+    else:
```
Comment (dgoldstein1) [inline backend/src/app/auth.py:14]: can you add a test for when a user isn't authenticated?

File: backend/src/services/user.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    self.db_session = Session()
+
+  def add_user(self, username: str, password: str, role: str, is_admin: bool = False) -> User:
+    _new_user = User(id = uuid.uuid4(), username = username, password_hash = pbkdf2_sha256.hash(password), role = role, role_id = uuid.uuid4(), is_admin = is_admin)
+    self.db_session.add(_new_user)
+    self.db_session.commit()
+    return _new_user
+
+  def get_user_by_id(self, uid) -> User:
+    return self.db_session.query(User).filter(User.id == uid).one_or_none()
```
Comment (dgoldstein1) [inline backend/src/services/user.py:23]: can you add tests here?

File: backend/src/services/auth.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+      spl = auth_header.split(" ")
+      if len(spl) != 2:
+          return (
+              None,
+              "Token header must be in form: 'Authorization: Bearer $JWT_TOKEN but was: '{}'"
+              .format(auth_header))
+      return app.services.auth_service().decode_jwt(spl[1])
+    except:
+      app.logger.error('Exception authenticating request')
+      return (None, "Exception authenticating request")
```
Comment (dgoldstein1) [inline backend/src/services/auth.py:87]: can you add tests for this file?

File: backend/src/auth_handlers.py [hunk 11]
```diff
@@ -1,131 +1,63 @@
...
-
+        user_data = request.get_json() #dictionary of input
+        username = user_data.get('user_name')
+        password = user_data.get('password')
+        role = user_data.get('role')
+        is_admin = bool(user_data.get('is_admin'))
+        if username and password and role and is_admin:
+            created_user = app.services.user_service().add_user(username,password,role,is_admin)
+            return jsonify({'jwt': app.services.auth_service().encode_jwt(created_user)})
+        else:
```
Comment (dgoldstein1) [inline backend/src/auth_handlers.py:23]: can you add tests for `/register`? Somehow we lost these

**Comment by thinkharderdev** [comment]
> ```shell
> curl -XPOST -H "content-type: application/json" -d '{"user_name" : "david", "password" : "davidrulz", "role_id" : "4", "role" : "navigator", "is_admin": true}' http://localhost:5000/auth/register
> ```

Are you still getting this. It is working for me. 

```shell 
curl -XPOST -H "content-type: application/json" -d '{"user_name" : "david", "password" : "davidrulz", "role_id" : "4", "role" : "navigator", "is_admin": true}' http://localhost:5000/auth/register
{
  "jwt": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1ODY2MjUxMDAsInVpZCI6ImQzNjc2ZDU5LWJkNmQtNGVmOS1hMWM2LTQ4Yjk3ZTMyMzUxMCIsInJvbGUiOiJuYXZpZ2F0b3IifQ.nlLbFBEqZYP939kf7LvvwbCJVTqzwS5hTHNcLqe0VVM"
}
```

File: backend/README.md (hunk 2 above)
Comment (thinkharderdev) [reply backend/README.md:53]: Weird, works for me. I added instructions for both ways in the README now

File: backend/api/swagger.yml (hunk 3 above)
Comment (thinkharderdev) [reply backend/api/swagger.yml]: Right, sorry that was accidental. Restores

File: backend/README.md (hunk 1 above)
Comment (thinkharderdev) [reply backend/README.md:12]: Sorry, accident

File: backend/src/auth_handlers.py (hunk 4 above)
Comment (thinkharderdev) [reply backend/src/auth_handlers.py]: Yep, done

File: backend/watch_dev_changes.sh (hunk 7 above)
Comment (thinkharderdev) [reply backend/watch_dev_changes.sh]: Done

File: backend/src/app/auth.py (hunk 8 above)
Comment (thinkharderdev) [reply backend/src/app/auth.py:14]: Adding tests in a separate PR

File: backend/src/services/user.py (hunk 9 above)
Comment (thinkharderdev) [reply backend/src/services/user.py:23]: Adding tests in a separate PR

File: backend/src/services/auth.py (hunk 10 above)
Comment (thinkharderdev) [reply backend/src/services/auth.py:87]: Adding tests in a separate PR

File: backend/src/auth_handlers.py (hunk 11 above)
Comment (thinkharderdev) [reply backend/src/auth_handlers.py]: Done. Good catch, I actually found a bug while adding back the unit test. Unit testing FTW!

File: backend/src/auth_handlers.py (hunk 5 above)
Comment (thinkharderdev) [reply backend/src/auth_handlers.py]: Done. I've been coding in Scala for a while where the use of return statements is generally discouraged. It still feels weird to use them :)

**Comment by dgoldstein1** [comment]
@thinkharderdev looks great! I am able to make a user now. I think my psql must have cached some data (?) because it works when I drop all the tables and reset everything.

Can you merge your branch with `code-for-charlottesville:master`? We should be able to test this with the UI. I think the only change we should need is changing `username` to `user_name` in the UI.
